src/GafferScene/Constraint.cpp

```diff
@@ -38,11 +38,594 @@
 
 #include "Gaffer/StringPlug.h"
 
+#include "IECore/CircularIterator.h"
+
+#include "IECoreScene/CurvesPrimitive.h"
+#include "IECoreScene/MeshPrimitive.h"
+#include "IECoreScene/PointsPrimitive.h"
+
+#include <OpenEXR/ImathMatrixAlgo.h>
+
+#include <boost/format.hpp>
+#include <boost/intrusive/list.hpp>
+
+#include <algorithm>
+#include <cassert>
+#include <cmath>
+#include <limits>
+
 using namespace Imath;
 using namespace IECore;
 using namespace Gaffer;
 using namespace GafferScene;
 
+namespace
+{
+
+template< typename Indexer, typename T >
+bool ptInPolygon( const Indexer indexer, const Imath::Vec2< T >& v, const int n )
+{
+	bool result = false;
+	for( int i = 0; i < n; ++i )
+	{
+		const Imath::Vec2< T >& v0 = indexer( i );
+		const Imath::Vec2< T >& v1 = indexer( ( i + 1 ) % n );
+
+		// NOTE: Algorithm 3.5, "Collision Detection in Interactive 3d Environments", Gino Van Den Bergen
+		//       with modifications to include end points and points on axis aligned edges.
+
+		if( v0 == v )
+		{
+			return true; // NOTE : end point
+		}
+		else if(
+			( ( v0.y == v1.y ) && ( v.y == v0.y ) && ( ( v0.x >= v.x ) != ( v1.x >= v.x ) ) ) ||
+			( ( v0.x == v1.x ) && ( v.x == v0.x ) && ( ( v0.y >= v.y ) != ( v1.y >= v.y ) ) ) )
+		{
+			return true; // NOTE : axis aligned edge
+		}
+		else if( ( v0.y >= v.y ) != ( v1.y >= v.y ) )
+		{
+			// Edge crosses horizontal line `y == v.y`
+
+			if( ( v0.x >= v.x ) != ( v1.x >= v.x ) )
+			{
+				// Edge crosses vertical line `x == v.x`
+
+				if( v0.x + ( v.y - v0.y ) * ( v1.x - v0.x ) / ( v1.y - v0.y ) >= v.x )
+				{
+					// Edge crosses the ray `y == v.y, x >= v.x`
+
+					result = !result;
+				}
+			}
+			else if( v0.x >= v.x )
+			{
+				// Edge crosses the ray `y == v.y, x >= v.x`
+
+				result = !result;
+			}
+		}
+	}
+
+	return result;
+}
+
+void constructLocalFrame( Imath::M44f& m, const Imath::V3f& p, const Imath::V3f& t, const Imath::V3f& b )
+{
+	// ignore tangents with zero length to avoid local frame collapse
+
+	const float threshold = 2.f * std::numeric_limits< float >::min();
+
+	if(
+		( t.length2() <= threshold ) ||
+		( b.length2() <= threshold ) )
+	{
+		m.translate( p );
+	}
+	else
+	{
+		m = Imath::computeLocalFrame( p, t, b );
+	}
+}
+
+struct UVIndexer
+{
+	UVIndexer( const IECoreScene::MeshPrimitive& primitive, const std::string& uvSet, const bool throwOnError )
+	: m_indices( 0 )
+	, m_uvIndices( 0 )
+	, m_uvs( 0 )
+	{
+		const IECoreScene::PrimitiveVariableMap::const_iterator uvIt = primitive.variables.find( uvSet );
```

Comment:
The main reason I did not use ```primitive.variableIndexedView( uvSet )``` is to avoid the unneccessary creation (allocation and copying) of the uv indices

Reply:
IndexedView should be a non-owning view onto data that exists already (or doesn't exist, since indices are optional). It shouldn't cause allocation or copying of anything - are you seeing otherwise?